### PR TITLE
Add Total Supply tracking

### PR DESCRIPTION
**Pull request by sdlyy**
Resolves L2B-1794

## Context
Data access layer for total supply entries

### CODE REVIEW

File: packages/backend/src/peripherals/database/migrations/065_total_supply.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+This is a migration file. Once created the file should not be renamed or edited,
+because migrations are only run once on the production server. 
+
+If you find that something was incorrectly set up in the `up` function you
+should create a new migration file that fixes the issue.
+
+*/
+
+import { Knex } from 'knex'
```
Comment (sdlyy) [inline packages/backend/src/peripherals/database/migrations/065_total_supply.ts:14]: Double check schema and indices please

File: packages/backend/src/peripherals/database/shared/types.ts [hunk 2]
```diff
@@ -103,6 +103,19 @@ declare module 'knex/types/tables' {
...
+    asset_id: string
+    total_supply: string
+    chain_id: number
+  }
+
+  interface TotalSupplyStatusRow {
+    chain_id: number
+    config_hash: string
+    unix_timestamp: Date
+  }
```
Comment (sdlyy) [inline packages/backend/src/peripherals/database/shared/types.ts:117]: Schema check please

File: packages/backend/src/peripherals/database/migrations/065_total_supply.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    table.index(['chain_id', 'config_hash'])
+
+    table.primary(['chain_id', 'unix_timestamp'])
+  })
+}
+
+export async function down(knex: Knex) {
+  await knex.schema.dropTable('total_supplies')
+  await knex.schema.dropTable('total_supplies_status')
+}
```
Comment (sdlyy) [inline packages/backend/src/peripherals/database/migrations/065_total_supply.ts:40]: Sanity double check please schema and indices based on usage

File: packages/backend/src/peripherals/database/shared/types.ts [hunk 4]
```diff
@@ -103,6 +103,19 @@ declare module 'knex/types/tables' {
...
+
```
Comment (sdlyy) [inline packages/backend/src/peripherals/database/shared/types.ts:118]: Also double check it here whether we agree on such row shape

File: packages/backend/src/peripherals/database/TotalSupplyRepository.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import { Database } from './shared/Database'
+
+export interface TotalSupplyRecord {
+  timestamp: UnixTime
+  totalSupply: bigint
+  assetId: AssetId
+  chainId: ChainId
+}
+
+export interface DataBoundary {
```
Comment (mateuszradomski) [inline packages/backend/src/peripherals/database/TotalSupplyRepository.ts:15]: Copy-pasta, this is not used anywhere

**Review by mateuszradomski — CHANGES_REQUESTED** [review]

File: packages/backend/src/peripherals/database/migrations/065_total_supply.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import { Knex } from 'knex'
+
+export async function up(knex: Knex) {
+  await knex.schema.createTable('total_supplies', function (table) {
+    table.dateTime('unix_timestamp').notNullable()
+    table.decimal('total_supply', 80, 0).notNullable()
+    table.string('asset_id').notNullable()
+    table.integer('chain_id').notNullable()
+
+    table.primary(['unix_timestamp', 'asset_id', 'chain_id'])
```
Comment (antooni) [inline packages/backend/src/peripherals/database/migrations/065_total_supply.ts:23]: change the order to [chain_id, unix_timestamp, asset_id]

**Review by antooni — CHANGES_REQUESTED** [review]
Please update the index, when the table will include mulitiple chainIds it would be better to first filter by them in the index

File: packages/backend/src/peripherals/database/TotalSupplyRepository.ts (hunk 5 above)
Comment (sdlyy) [reply packages/backend/src/peripherals/database/TotalSupplyRepository.ts]: https://github.com/l2beat/l2beat/pull/1663/commits/f38d327505143c45f44435ca720343899f7683a5

File: packages/backend/src/peripherals/database/migrations/065_total_supply.ts (hunk 6 above)
Comment (sdlyy) [reply packages/backend/src/peripherals/database/migrations/065_total_supply.ts]: https://github.com/l2beat/l2beat/pull/1663/commits/80edc53aee622756e426907bbe0ef571f70aea99

**Review by antooni — APPROVED** [review]

**Review by mateuszradomski — APPROVED** [review]

**Pull request merged**